commons/src/main/scala/com/lynbrookrobotics/potassium/commons/drivetrain/PurePursuitControllers.scala

```diff
@@ -165,5 +163,70 @@ trait PurePursuitControllers {
       }
     )
   }
+
+  def followWayPointsController(wayPoints: Seq[Point], position: PeriodicSignal[Point])
+                               (implicit hardware: UnicycleHardware,
+                                signal: Signal[UnicycleProperties]): (PeriodicSignal[UnicycleSignal], Signal[Option[Length]]) = {
+    var previousLookAheadPoint: Option[Point] = None
+
+    val biSegmentPaths = wayPoints.sliding(3).map { points =>
+      (
+        Segment(points(0), points(1)),
+        if (points.size == 3) {
+          Some(Segment(points(1), points(2)))
+        } else {
+          None
+        }
+      )
+    }
+
+    var currPath: (Segment, Option[Segment]) = (
+      Segment(wayPoints(0), wayPoints(1)),
+      if (wayPoints.size >= 3) {
+        Some(Segment(wayPoints(1), wayPoints(2)))
+      } else {
+        None
+      }
+    )
+
+    val selectedPath = Signal {
+      if (previousLookAheadPoint.forall(p => currPath._2.get.containsInXY(p, Feet(0.1)))) {
+        if (biSegmentPaths.hasNext) {
+          println("**********advancing path ***********")
+          currPath = biSegmentPaths.next()
+        }
+      }
+
+      currPath
+    }
+
+    val initialTurnPosition = hardware.turnPosition.get
+
+    val (turnOutput, heading, lookAheadPoint) = purePursuitControllerTurn(
+      hardware.turnPosition.map(_ - initialTurnPosition),
+      position,
+      selectedPath)
+    val historyUpdatingLookAheadPoint = lookAheadPoint.withCheck{p =>
+      previousLookAheadPoint = Some(p)
+    }
+
+    val forwardOutput = pointDistanceControl(
+      position,
+      historyUpdatingLookAheadPoint)._1
+    val distanceToLast = position.map(_ distanceTo wayPoints.last)
+
+    val errorToLast = distanceToLast.peek.map { d =>
```

Comment:
This breaks  down if after traveling some path, we need to return to the starting point. Instead use the distance to the look ahead point